File: Tiny2D/Java/src/net/sapium/tiny/Game.java
package net.sapium.tiny;

import java.awt.Canvas;
import java.awt.Color;
import java.awt.Component;
import java.awt.Dimension;
import java.awt.Frame;
import java.awt.Graphics2D;
import java.awt.event.WindowEvent;
import java.awt.event.WindowListener;
import java.awt.image.BufferStrategy;
import java.io.IOException;

import javax.imageio.ImageIO;

import net.sapium.tiny.graphics.Screen;
import net.sapium.tiny.utils.InputHandler;

import org.apache.log4j.ConsoleAppender;
import org.apache.log4j.FileAppender;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.log4j.PatternLayout;

public class Game implements Runnable {
    private String title = "Game";
    private int width = 720;
    private int height = 480;
    private int fps = 60;
    private int tps = 60;
    private Color background;
    private boolean printFPS = true;
    private boolean limitFPS = true;
    private boolean limitTPS = true;

    private String icon = "";
    private String logfile = "output.log";
    
    private Canvas canvas;
    
    private Logger logger = Logger.getLogger(Game.class);
    private Thread thread;
    private boolean running = false;
    private Screen currentScreen;
    
    public Game() {
        canvas = new Canvas();
        canvas.setPreferredSize(new Dimension(width, height));
    }
    
    public Game(String title, int width, int height) {
        this.title = title;
        this.width = width;
        this.height = height;
        
        canvas = new Canvas();
        canvas.setPreferredSize(new Dimension(width, height));
    }

    /**
     * Draws to the screen
     */
    private void render() {
        BufferStrategy bs = canvas.getBufferStrategy();
        if(bs == null){
            canvas.createBufferStrategy(2);
            return;
        }
        
        Graphics2D g = (Graphics2D) bs.getDrawGraphics();
        
        g.clearRect(0, 0, width, height);
        
        currentScreen.draw(g);
        
        g.dispose();
        bs.show();
    }

    /**
     * Processes game updates
     */
    private void tick() {
        currentScreen.tick();
    }

    /**
     * Calls tick, then render at the specified rates
     */
    @Override
    public void run() {
        double unprocessedTicks = 0.0;
        double nsPerTick = 1000000000.0/tps;
        
        double unrenderedFrames = 0.0;
        double nsPerFrame = 1000000000.0/fps;

        long lastTime = System.nanoTime();
        
        long lastFpsTime = System.currentTimeMillis();
        int frameCount = 0;
        int tickCount = 0;
        
        while(running){
            long now = System.nanoTime();
            unprocessedTicks += (now - lastTime) / nsPerTick;
            unrenderedFrames += (now - lastTime) / nsPerFrame;
            lastTime = now;
            
            if(unprocessedTicks >= 1.0 || !limitTPS){
                tick();
                unprocessedTicks -= 1.0;
                tickCount++;
            }
            
            if(unrenderedFrames >= 1.0 || !limitFPS){
                render();
                unrenderedFrames -= 1.0;
                frameCount++;
            }
            
            if(printFPS && System.currentTimeMillis() - lastFpsTime > 1000){
                System.out.println(frameCount + " fps; " + tickCount + " tps");
                frameCount = 0;
                tickCount = 0;
                lastFpsTime = System.currentTimeMillis();
            }
        }
    }
    
    /**
     * Set the window title
     * 
     * must be set before init() is called
     * 
     * Default: "Game"
     * 
     * @param title the window title
     */
    public void setTitle(String title) {
        this.title = title;
    }
    
    /**
     * Set the location of the window icon
     * 
     * must be called before createWindow()
     * 
     * if null don't set an icon
     * 
     * Default: /icon.png
     * 
     * @param path location of the icon image
     */
    public void setIcon(String path) {
        this.icon = path;
    }
    
    /**
     * Sets the frames per second to limit rendering to
     * 
     * if <= 0 unlimited
     * 
     * Default: 60
     * 
     * @param fps the fps to set
     */
    public void setFps(int fps) {
        this.fps = fps;
        
        if(fps <= 0)
            this.limitFPS = false;
        else
            this.limitFPS = true;
    }
    
    /**
     * Sets the number of updates per second to limit to
     * 
     * if <= 0 unlimited
     * 
     * Default: 60
     * 
     * @param tps the tps to set
     */
    public void setTPS(int tps) {
       this.tps = tps;
       
       if(tps <= 0) 
           this.limitTPS = false;
       else
           this.limitTPS = true;
    }

    /**
     * Sets the background color to use when clearing the screen
     * 
     * @param background set the background color to draw
     */
    public void setBackgroundColor(Color background) {
        this.background = background;
        canvas.setBackground(background);
    }

    /**
     * Sets whether or not the engine should print the current fps
     * 
     * Default: true
     * 
     * @param printFPS if true print the fps and tps on stdout
     */
    public void setPrintFPS(boolean printFPS) {
        this.printFPS = printFPS;
    }

    /**
     * Sets the screen to draw
     * 
     * @param s the screen to draw
     */
    public void setCurrentScreen(Screen s){
        currentScreen = s;
    }
    
    /**
     * Set the file Log4J should append to
     * 
     * must be called before init
     * 
     * if null, only log to the console (unnecessary if you're running in an applet)
     * 
     * @param logfile filename to use
     */
    public void setLogfile(String logfile) {
        this.logfile = logfile;
    }
    
    /**
     * @return The component all of the rendering is done onto (add this to your window or applet)
     */
    public Component getDrawComponent() {
        return canvas;
    }
    
    /**
     * Sets log4J to print to the console and to the logfile if set
     * 
     * if running in an applet it does not write to a logfile due to security restrictions
     */
    private void initLog4J() {
        logger.setLevel(Level.INFO);
        
        Logger.getRootLogger().addAppender(new ConsoleAppender(new PatternLayout("%d %-5p [%c] - %m%n")));
        try {
            if(logfile != null)
                Logger.getRootLogger().addAppender(new FileAppender(new PatternLayout("%d %-5p [%c] - %m%n"), logfile));
        } catch (IOException e) {
            logger.error("Could not write log file", e);
        } catch (SecurityException e) {
            logger.debug("Could not create log file. Running in an applet.");
        }
    }
    
    /**
     * Creates a window and initializes the game
     */
    public void createWindow() {
        Frame window = new Frame(this.title);
        
        final Game game = this;
        window.addWindowListener(new WindowListener() {
            @Override
            public void windowClosing(WindowEvent e) {
                game.stop();
                System.exit(0);
            }
            @Override
            public void windowOpened(WindowEvent e) {}
            @Override
            public void windowClosed(WindowEvent e) {}
            @Override
            public void windowIconified(WindowEvent e) {}
            @Override
            public void windowDeiconified(WindowEvent e) {}
            @Override
            public void windowActivated(WindowEvent e) {}
            @Override
            public void windowDeactivated(WindowEvent e) {}
        });
        
        initLog4J();
        
        this.init();
        window.add(canvas);
        
        if(icon != null && icon != "") {
            try {
                window.setIconImage(ImageIO.read(Game.class.getResourceAsStream(icon)));
            } catch (IOException e) {
                logger.error("Could not load icon image", e);
            }
        }

        window.setLocationRelativeTo(null);
        window.setResizable(false);
        window.pack();
        window.setVisible(true);
    }
    
    /**
     * Initializes the game
     */
    public void init() {
        initLog4J();
        
        InputHandler ih = new InputHandler();
        canvas.addKeyListener(ih);
        canvas.addMouseListener(ih);
        canvas.addMouseMotionListener(ih);
        
        logger.debug("init");
        canvas.setBackground(background);
        thread = new Thread(this);
    }

    /**
     * Starts running the game
     */
    public void start() {
        logger.debug("start");
        running = true;
        thread.start();
    }

    /**
     * Stops the game
     */
    public void stop() {
        logger.debug("stop");
        running = false;
    }
}

File: Tiny2D/C++/lib/Game.cpp
#include <stdio.h>
#include "SDL_image.h"
#include "Game.h"
#include "utils/Logger.h"

using namespace Tiny2D;

Logger logger("Game");

Game::Game(InputHandler *ih, const char *title, int width, int height) :
    ih(ih), title(title), width(width), height(height) {
    ih->setOnQuitCallback(&quit, this);
}

Game::~Game() {
    stop();
}

void Game::tick(unsigned int dt) {
    ih->processEventQueue();

    if(this->currentScreen != nullptr)
        this->currentScreen->tick(dt);
}

void Game::render() {
    SDL_FillRect(this->surface, nullptr, this->background);

    if(this->currentScreen != nullptr)
        this->currentScreen->draw(this->surface);

    SDL_UpdateWindowSurface(this->window);
}

int Game::run(void *data) {
    Game *g = (Game*) data;

    double unprocessedTicks = 0.0;
    double msPerTick = 1000.0/g->tps;

    double unrenderedFrames = 0.0;
    double msPerFrame = 1000.0/g->fps;

    unsigned int lastUpdateTime = SDL_GetTicks();
    unsigned int lastTickTime = SDL_GetTicks();;

    unsigned int lastFpsTime = SDL_GetTicks();
    int frameCount = 0;
    int tickCount = 0;

    while(g->running) {
        if(g->changedRates) {
            if(g->limitTps) {
                msPerTick = 1000.0/g->tps;
                unprocessedTicks = 0.0;
            }
            if(g->limitFps) {
                msPerFrame = 1000.0/g->fps;
                unrenderedFrames = 0.0;
            }

            g->changedRates = false;
        }

        if(g->newScreen != nullptr) {
            g->currentScreen = g->newScreen;
            g->newScreen = nullptr;
        }

        unsigned int now = SDL_GetTicks();
        unprocessedTicks += (now - lastUpdateTime) / msPerTick;
        unrenderedFrames += (now - lastUpdateTime) / msPerFrame;
        lastUpdateTime = now;

        if(!g->limitTps || unprocessedTicks >= 1.0) {
            g->tick(now - lastTickTime);
            unprocessedTicks -= 1.0;
            tickCount++;
            //set to now instead of SDL_GetTicks() to count the time to process in the delta time
            lastTickTime = now;
        }

        if(!g->limitFps || unrenderedFrames >= 1.0) {
            g->render();
            unrenderedFrames -= 1.0;
            frameCount++;
        }

        if(g->printFps && SDL_GetTicks() - lastFpsTime > 1000) {
            //don't use logger for this so we don't clutter the log file
            printf("%d fps; %d tps\n", frameCount, tickCount);
            frameCount = tickCount = 0;
            lastFpsTime = SDL_GetTicks();
        }
    }
}

void Game::start() {
    createWindow();

    logger.debug("Starting game");
    running = true;
    this->thread = SDL_CreateThread(&Game::run, "Game", this);
}

void Game::stop() {
    running = false;

    SDL_DestroyWindow(this->window);
    this->window = nullptr;
    SDL_Quit();
}

void Game::waitThread() {
    SDL_WaitThread(this->thread, nullptr);
}

void Game::quit(void *g) {
    ((Game *)g)->running = false;
    logger.debug("Stopping game");
}

void Game::createWindow() {
    if(SDL_Init(SDL_INIT_VIDEO) < 0) {
        logger.error("SDL could not initialize");
        logger.error(SDL_GetError());
    }else{
        window = SDL_CreateWindow(this->title, SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED,
                                  this->width, this->height, SDL_WINDOW_SHOWN);

        if(this->window == nullptr) {
            logger.error("SDL could not create window");
            logger.error(SDL_GetError());
        }else{
            this->setIcon(this->icon);

            this->surface = SDL_GetWindowSurface(this->window);

            this->setBackgroundColor(this->background);
        }
    }
}

void Game::setIcon(const char *icon) {
    this->icon = icon;

    if(this->window != nullptr) {
        SDL_Surface *image = IMG_Load(icon);

        if(image != nullptr) {
            SDL_SetWindowIcon(this->window, image);
            SDL_FreeSurface(image);
        }
    }
}

void Game::setFps(int fps) {
    this->fps = fps;

    this->limitFps = fps <= 0 ? false : true;
    this->changedRates = true;
}

void Game::setTps(int tps) {
    this->tps = tps;

    this->limitTps = tps <= 0 ? false : true;
    this->changedRates = true;
}

void Game::setBackgroundColor(unsigned int color) {
    if(this->surface == nullptr) {
        this->background = color;
    }else{
        this->background = SDL_MapRGB(this->surface->format, (color >> 16) & 0xFF, (color >> 8) & 0xFF, color & 0xFF);
    }
}

void Game::setPrintFps(bool printFps) {
    this->printFps = printFps;
}

void Game::setCurrentScreen(Screen *screen) {
    this->newScreen = screen;
}
